Update internationalization enhancement for how to handle login template translation strings (#724)
openshift/enhancements: enhancements/console/internationalization.md

We will also modify our existing file conversion tools to handle translation files spread across multiple packages so we can easily export/import translation files.

> Translation of the login page will be handled by [oauth-server](https://github.com/openshift/oauth-server), which will check firstly the request URL for `ui_locales` query. If the query is missing, `oauth-server` will check for request `Accept-Language` header. If both the query and header are missing in the request, `oauth-server` will default to English language for localization.

Reviewer: ```suggestion
Translation of the login page will be handled by [oauth-server](https://github.com/openshift/oauth-server), which will check for request `Accept-Language` header. If the header is missing in the request, `oauth-server` will default to English language for localization.
```

> [oauth-server](https://github.com/openshift/oauth-server) contains the default OKD login templates. Custom templates are stored in the [oauth-templates](https://github.com/openshift/oauth-templates) repository and will also need to be updated for the other brands like the OCP and OpenShift Dedicated login pages.

Reviewer 2: do we need a bump of oauth-server when the templates change? I assume the console team will do those?

Reviewer: yes, if any of the templates change we will make needed changes.
Will incorporate that into the enhancement

Reviewer: In case of any changes in these repositories, Console-Team will be responsible for syncing those changes between them.

Author: updated text

We will also modify our existing file conversion tools to handle translation files spread across multiple packages so we can easily export/import translation files.

> Translation of the login page will be handled by [oauth-server](https://github.com/openshift/oauth-server), which will check firstly the request URL for `lng` query. If the query is missing, `oauth-server` will check for request `Accept-Language` header. If both the query and header are missing in the request, `oauth-server` will differ to english language for localization.

Reviewer 3: s/differ/default

Reviewer 2: Why `lng` query **and** header? Why is the latter not enough?

Reviewer 3: > Why `lng` query **and** header? Why is the latter not enough?

@sttts We need to be able to override the system setting since we have a user preference for selecting a language in console. If there is a standard query parameter, we can definitely use that instead.

I wouldn't drop the `Accept-Language` header because we want the default to be the system language when the user hasn't explicitly selected something different as a language preference in console.

Reviewer 4: > I wouldn't drop the Accept-Language header because we want the default to be the system language when the user hasn't explicitly selected something different as a language preference in console.

That's fair. I would also expect the `ui_locales` to configure the session if set but I see no cookie handling here. Also, I am missing a description of the UI element that allows configuring this on the oauth-server side.

Reviewer: Thats a good question. I think for now it has to be manual.

Reviewer 3: I think console would store the last selected language in browser `localStorage` and then add `ui_locales` to the request when redirecting the user to login page. I haven't looked closely at the spec, though. Is this a query parameter?

Reviewer: @spadgett thoughts?

> [oauth-server](https://github.com/openshift/oauth-server) contains default login templates. Custom templates are stored in [oauth-templates](https://github.com/openshift/oauth-templates) repository.

Reviewer 3: ```suggestion
[oauth-server](https://github.com/openshift/oauth-server) contains the default OKD login templates. Custom templates are stored in the [oauth-templates](https://github.com/openshift/oauth-templates) repository and will also need to be updated for the other brands like the OCP and OpenShift Dedicated login pages.
```

> Since there are only a small number of strings, that rarely change, the localized strings for each of the supported languages will be directly part of the `oauth-server`, so we don't have to load them during runtime.

Reviewer 3: I'm a little unclear what this is saying.

Reviewer: @spadgett it's referring to https://github.com/openshift/oauth-server/blob/60b81ecc54f02b6f2fd4e9bba36c06888bee0272/pkg/server/login/locales.go#L41-L50

Maybe to make it clear we could say:
```suggestion
Since there are only a small number of strings, that rarely change, the localized strings for each of the supported languages will be directly hardcoded into the `oauth-server`, so we don't have to load them during runtime.
```

Reviewer 3: We use https://www.i18next.com/, but this seems heavy weight for a handful of strings in a static HTML page. I'm not sure how this would work with login page customization as well, since those pages would need to include the script. It's something we could explore though.

Reviewer 4: What's the plan going forward, then? This only localizes the console, but the information about the cluster state (operator statuses, for example) are still going to be in English. Is that ok or do you expect these to be translated eventually, too?

If they _are_ going to be translated, we'd need a bit more modularity in those translations. But I guess for now we can go with what you're proposing.

Reviewer: Agreed, since there are just a handful of strings I would prefer to put them directly into a go pkg, as is currently proposed.

Reviewer 3: > What's the plan going forward, then? This only localizes the console, but the information about the cluster state (operator statuses, for example) are still going to be in English. Is that ok or do you expect these to be translated eventually, too?

We don't currently have any plan to translate data coming from k8s resources, only the messages in console itself.

We will also modify our existing file conversion tools to handle translation files spread across multiple packages so we can easily export/import translation files.

> Translation of the login templates within [oauth-templates](https://github.com/openshift/oauth-templates) and [oauth-server](https://github.com/openshift/oauth-server) repositories will be manually updated by the console team since their are only a small number of strings that rarely change. These strings are to be included in the console repository so that the i18next-parser will add them in the files sent to the Globalization team for translation.

Reviewer 3: We should add more detail here like how we are using the `Accept-Language` header and optional `lng` query parameter to determine what language to show on the backend. We might want @jhadvig to contribute that detail since he did the backend work. We should also discuss if/how we plan to handle the language user preference for the login page. (See discussion in https://github.com/openshift/oauth-server/pull/71)

Author: @spadgett The text has been updated with additional detail provided by @jhadvig
